File: typescript/packages/x402/src/core/client/balance-check.ts
/**
 * Client-side pre-flight balance check.
 *
 * Wires a `BeforePaymentCreationHook` that queries the configured CDP wallet's
 * token balance for the requested asset/network and fails fast with a clear
 * {@link InsufficientFundsError} when the balance is below the amount the
 * server is asking for.
 *
 * Without this check, an unfunded wallet submits a signed payment that the
 * facilitator/verifier then rejects with a generic
 * `invalidReason: invalid_payload` + `execution reverted` message, which is
 * not actionable. This pre-check converts that into a typed error before any
 * signing occurs.
 *
 * Balance lookup is best-effort and is silently skipped when:
 * - the requested network is not supported.
 *   EVM coverage: `base` and `base-sepolia` use the CDP indexed token-balance
 *   API; `polygon`, `arbitrum`, `world`, and `world-sepolia` fall back to a
 *   direct on-chain `balanceOf` call via the public RPC endpoint from
 *   {@link CDP_EVM_RPC_URLS}. SVM coverage is `solana` and `solana-devnet`.
 * - the requested asset contract / mint is not present in the returned token
 *   balances (e.g. brand-new ERC-20 the CDP indexer hasn't seen yet), or
 * - the balance call itself fails (network blip, auth issue) — the original
 *   downstream error path is preserved.
 *
 * @packageDocumentation
 */
import type { CdpClient } from "@coinbase/cdp-sdk";
import type { BeforePaymentCreationHook } from "@x402/core/client";
import { createPublicClient, http } from "viem";

import {
  CDP_EVM_RPC_URLS,
  baseMainnetCaip2,
  baseSepoliaCaip2,
  solanaDevnetCaip2,
  solanaMainnetCaip2,
} from "../facilitator/constants.js";
import { tryParseAtomicFromRequirement } from "../guardrails/apply.js";
import { normalizeAsset } from "../guardrails/normalize.js";

/**
 * CAIP-2 → CDP indexed balance API network slug for CDP supported EVM networks.
 */
const CAIP_TO_CDP_EVM_BALANCE_NETWORK: Record<string, "base" | "base-sepolia"> = {
  [baseMainnetCaip2]: "base",
  [baseSepoliaCaip2]: "base-sepolia",
};

/** Minimal ERC-20 ABI fragment for `balanceOf`. */
const ERC20_BALANCE_OF_ABI = [
  {
    name: "balanceOf",
    type: "function",
    inputs: [{ name: "owner", type: "address" }],
    outputs: [{ name: "", type: "uint256" }],
    stateMutability: "view",
  },
] as const;

/**
 * CAIP-2 → CDP Solana balance API network.
 */
const CAIP_TO_CDP_SVM_BALANCE_NETWORK: Record<string, "solana" | "solana-devnet"> = {
  [solanaMainnetCaip2]: "solana",
  [solanaDevnetCaip2]: "solana-devnet",
};

/**
 * Thrown when the configured wallet's balance is below the amount required by
 * the server's `PaymentRequirements`.
 */
export class InsufficientFundsError extends Error {
  readonly code = "insufficient_funds" as const;
  readonly required: bigint;
  readonly available: bigint;
  readonly asset: string;
  readonly network: string;
  readonly address: string;

  constructor(params: {
    required: bigint;
    available: bigint;
    asset: string;
    network: string;
    address: string;
  }) {
    super(
      `Insufficient funds for x402 payment: wallet ${params.address} on ${params.network} has ` +
        `${params.available} of asset ${params.asset}, but ${params.required} is required.`,
    );
    this.name = "InsufficientFundsError";
    this.required = params.required;
    this.available = params.available;
    this.asset = params.asset;
    this.network = params.network;
    this.address = params.address;
  }
}

interface BalanceCheckOptions {
  cdpClient: CdpClient;
  evmAddress: `0x${string}`;
  svmAddress: string;
  /**
   * Optional warning sink for skipped-check diagnostics. Defaults to
   * `console.warn`. Pass a no-op to silence completely.
   */
  onWarning?: (message: string, cause?: unknown) => void;
  /**
   * Override the public JSON-RPC endpoints used for on-chain balance lookups,
   * keyed by CAIP-2 network identifier. Merged over the built-in
   * `CDP_EVM_RPC_URLS` defaults — only supply the networks you want to
   * override.
   */
  rpcUrls?: Partial<Record<string, { rpcUrl: string }>>;
}

const defaultWarning = (message: string, cause?: unknown): void => {
  if (cause === undefined) console.warn(`[@coinbase/x402] ${message}`);
  else console.warn(`[@coinbase/x402] ${message}`, cause);
};

/**
 * Builds a `BeforePaymentCreationHook` that performs a pre-flight balance check
 * against the configured CDP wallet.
 *
 * The hook is intentionally lenient: any condition under which the check
 * cannot be performed (unsupported network, missing token in balances, API
 * error) is treated as a pass so we never block legitimate payments on
 * pre-check noise.
 */
export function createBalanceCheckHook(opts: BalanceCheckOptions): BeforePaymentCreationHook {
  const { cdpClient, evmAddress, svmAddress } = opts;
  const warn = opts.onWarning ?? defaultWarning;
  const rpcUrls: Record<string, { rpcUrl: string }> = opts.rpcUrls
    ? ({ ...CDP_EVM_RPC_URLS, ...opts.rpcUrls } as Record<string, { rpcUrl: string }>)
    : CDP_EVM_RPC_URLS;

  return async (context) => {
    const req = context.selectedRequirements;
    const required = tryParseAtomicFromRequirement(req);
    if (required === undefined) return undefined;
    if (required === 0n) return undefined;

    const normalizedAsset = normalizeAsset(req.asset);

    const evmNetwork = CAIP_TO_CDP_EVM_BALANCE_NETWORK[req.network];
    const evmRpcUrl = rpcUrls[req.network]?.rpcUrl;
    const svmNetwork = CAIP_TO_CDP_SVM_BALANCE_NETWORK[req.network];

    let available: bigint;
    let address: string;

    const assetLabel = svmNetwork ? req.asset : normalizedAsset;

    try {
      if (evmNetwork) {
        address = evmAddress;
        available = await getEvmBalance(cdpClient, {
          address: evmAddress,
          network: evmNetwork,
          asset: normalizedAsset,
        });
      } else if (evmRpcUrl) {
        if (!normalizedAsset.startsWith("0x")) return undefined;
        address = evmAddress;
        available = await readOnChainErc20Balance(
          evmRpcUrl,
          normalizedAsset as `0x${string}`,
          evmAddress,
        );
      } else if (svmNetwork) {
        address = svmAddress;
        available = await getSvmBalance(cdpClient, {
          address: svmAddress,
          network: svmNetwork,
          asset: req.asset,
        });
      } else {
        return undefined;
      }
    } catch (error) {
      warn(
        `Pre-flight balance check failed for ${req.network} (${assetLabel}); ` +
          "proceeding without check.",
        error,
      );
      return undefined;
    }

    if (available >= required) return undefined;

    throw new InsufficientFundsError({
      required,
      available,
      asset: req.asset,
      network: req.network,
      address,
    });
  };
}

/**
 * Reads the ERC-20 `balanceOf` for `walletAddress` directly on-chain via the
 * supplied public JSON-RPC endpoint. Used for CDP facilitator networks that are
 * not covered by the CDP indexed token-balance API.
 */
async function readOnChainErc20Balance(
  rpcUrl: string,
  contractAddress: `0x${string}`,
  walletAddress: `0x${string}`,
): Promise<bigint> {
  const client = createPublicClient({ transport: http(rpcUrl) });
  return client.readContract({
    address: contractAddress,
    abi: ERC20_BALANCE_OF_ABI,
    functionName: "balanceOf",
    args: [walletAddress],
  });
}

/**
 * Returns the balance (in atomic units) of `params.asset` held by
 * `params.address`, or `0n` when the asset is absent.
 */
async function getEvmBalance(
  cdpClient: CdpClient,
  params: { address: `0x${string}`; network: "base" | "base-sepolia"; asset: string },
): Promise<bigint> {
  let pageToken: string | undefined;
  do {
    const page = await cdpClient.evm.listTokenBalances({
      address: params.address,
      network: params.network,
      pageToken,
    });
    for (const entry of page.balances) {
      if (normalizeAsset(entry.token.contractAddress) === params.asset) {
        return entry.amount.amount;
      }
    }
    pageToken = page.nextPageToken;
  } while (pageToken);
  return 0n;
}

/**
 * Returns the balance (in atomic units) of `params.asset` held by
 * `params.address`, or `0n` when the asset is absent.
 */
async function getSvmBalance(
  cdpClient: CdpClient,
  params: { address: string; network: "solana" | "solana-devnet"; asset: string },
): Promise<bigint> {
  let pageToken: string | undefined;
  do {
    const page = await cdpClient.solana.listTokenBalances({
      address: params.address,
      network: params.network,
      pageToken,
    });
    for (const entry of page.balances) {
      if (entry.token.mintAddress === params.asset) {
        return entry.amount.amount;
      }
    }
    pageToken = page.nextPageToken;
  } while (pageToken);
  return 0n;
}

File: python/x402/cdp_x402/core/balance_check.py
"""Client-side pre-flight balance check.

Wires a ``BeforePaymentCreationHook`` that queries the configured CDP wallet's
token balance for the requested asset/network and fails fast with a clear
:exc:`InsufficientFundsError` when the balance is below the amount the server
is asking for.

Without this check, an unfunded wallet submits a signed payment that the
facilitator/verifier rejects with a generic
``invalidReason: invalid_payload`` + ``execution reverted`` message, which is
not actionable. This pre-check converts that into a typed error before any
signing occurs.

Balance lookup is best-effort and is silently skipped when:

- the requested network is not supported.
  EVM coverage: ``base`` and ``base-sepolia`` use the CDP indexed
  token-balance API; ``polygon``, ``arbitrum``, ``world``, and
  ``world-sepolia`` fall back to a direct on-chain ``balanceOf`` call via
  the public RPC endpoint from :data:`~cdp_x402.core.constants.CDP_EVM_RPC_URLS`.
  SVM coverage is ``solana`` and ``solana-devnet``.
- the requested asset contract / mint is not present in the returned token
  balances (e.g. a brand-new ERC-20 the CDP indexer has not seen yet), or
- the balance call itself fails (network blip, auth issue) — the original
  downstream error path is preserved.
"""

from __future__ import annotations

from collections.abc import Callable, Coroutine
from typing import TYPE_CHECKING, Any

from web3 import AsyncWeb3
from web3.providers import AsyncHTTPProvider

from cdp_x402.core.constants import (
    BASE_MAINNET_CAIP2,
    BASE_SEPOLIA_CAIP2,
    CDP_EVM_RPC_URLS,
    SOLANA_DEVNET_CAIP2,
    SOLANA_MAINNET_CAIP2,
)
from cdp_x402.core.guardrails.normalize import normalize_asset
from cdp_x402.core.utils import default_warn, read_required_atomic

if TYPE_CHECKING:
    from cdp import CdpClient

# ---------------------------------------------------------------------------
# Network maps — only networks the CDP balance APIs support today
# ---------------------------------------------------------------------------

# CAIP-2 → CDP indexed balance API network slug.
# The CDP list_token_balances API only supports a subset of networks; for the
# remaining CDP facilitator EVM networks we fall back to a direct on-chain
# balanceOf call using CDP_EVM_RPC_URLS.
_CAIP_TO_CDP_EVM_NETWORK: dict[str, str] = {
    BASE_MAINNET_CAIP2: "base",
    BASE_SEPOLIA_CAIP2: "base-sepolia",
}

# CAIP-2 → CDP Solana balance API network string.
_CAIP_TO_CDP_SVM_NETWORK: dict[str, str] = {
    SOLANA_MAINNET_CAIP2: "solana",
    SOLANA_DEVNET_CAIP2: "solana-devnet",
}


# ---------------------------------------------------------------------------
# Error
# ---------------------------------------------------------------------------


class InsufficientFundsError(Exception):
    """Raised when the wallet's balance is below the required payment amount.

    The ``code`` is fixed to ``"insufficient_funds"`` so callers can branch on
    it regardless of the human-readable message.
    """

    code: str = "insufficient_funds"

    def __init__(
        self,
        *,
        required: int,
        available: int,
        asset: str,
        network: str,
        address: str,
    ) -> None:
        super().__init__(
            f"Insufficient funds for x402 payment: wallet {address} on {network} has "
            f"{available} of asset {asset}, but {required} is required."
        )
        self.required = required
        self.available = available
        self.asset = asset
        self.network = network
        self.address = address


# ---------------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------------


async def _get_evm_balance(
    cdp_client: Any,
    address: str,
    network: str,
    asset: str,
) -> int:
    """Return the EVM token balance for ``asset`` at ``address``.

    Paginates through CDP indexed balances and returns on the first matching
    entry. Returns ``0`` when the asset is absent (i.e. the wallet holds no
    tokens of that type).
    """
    page_token: str | None = None
    while True:
        page = await cdp_client.evm.list_token_balances(
            address=address,
            network=network,
            page_token=page_token,
        )
        for entry in page.balances:
            if normalize_asset(entry.token.contract_address) == asset:
                return int(entry.amount.amount)
        page_token = page.next_page_token
        if not page_token:
            break
    return 0


_ERC20_BALANCE_OF_ABI = [
    {
        "name": "balanceOf",
        "type": "function",
        "inputs": [{"name": "owner", "type": "address"}],
        "outputs": [{"name": "", "type": "uint256"}],
        "stateMutability": "view",
    }
]


async def _read_onchain_erc20_balance(
    rpc_url: str,
    token_address: str,
    wallet_address: str,
) -> int:
    """Query an ERC-20 ``balanceOf`` on-chain via ``AsyncWeb3``.

    Used for CDP facilitator EVM networks that are not covered by the CDP
    indexed token-balance API (Polygon, Arbitrum, World, World Sepolia).

    :param rpc_url: Public JSON-RPC endpoint for the target network.
    :param token_address: ERC-20 contract address.
    :param wallet_address: Wallet address to check the balance of.
    :returns: Token balance in atomic units.
    """
    w3 = AsyncWeb3(AsyncHTTPProvider(rpc_url))
    try:
        contract = w3.eth.contract(
            address=AsyncWeb3.to_checksum_address(token_address),
            abi=_ERC20_BALANCE_OF_ABI,
        )
        result: int = await contract.functions.balanceOf(
            AsyncWeb3.to_checksum_address(wallet_address)
        ).call()
        return result
    finally:
        await w3.provider.disconnect()


async def _get_svm_balance(
    cdp_client: Any,
    address: str,
    network: str,
    asset: str,
) -> int:
    """Return the Solana token balance for ``asset`` at ``address``.

    Paginates through CDP indexed balances and returns on the first matching
    entry. Returns ``0`` when the asset is absent.
    """
    page_token: str | None = None
    while True:
        page = await cdp_client.solana.list_token_balances(
            address=address,
            network=network,
            page_token=page_token,
        )
        for entry in page.balances:
            if entry.token.mint_address == asset:
                return int(entry.amount.amount)
        page_token = page.next_page_token
        if not page_token:
            break
    return 0


# ---------------------------------------------------------------------------
# Hook factory
# ---------------------------------------------------------------------------

_WarnFn = Callable[[str, object | None], None]


def create_balance_check_hook(
    *,
    cdp_client: CdpClient,
    evm_address: str,
    svm_address: str,
    on_warning: _WarnFn | None = None,
    rpc_urls: dict[str, str] | None = None,
) -> Callable[[Any], Coroutine[Any, Any, None]]:
    """Build a ``BeforePaymentCreationHook`` that performs a pre-flight balance check.

    The hook is intentionally lenient: any condition under which the check
    cannot be performed (unsupported network, missing token in balances, API
    error) is treated as a pass so legitimate payments are never blocked on
    pre-check noise.

    :param cdp_client: Authenticated :class:`cdp.CdpClient` instance.
    :param evm_address: The EVM wallet address to check balances against.
    :param svm_address: The Solana wallet address to check balances against.
    :param on_warning: Optional callable ``(message, cause)`` for skipped-check
        diagnostics. Defaults to :mod:`logging` at WARNING level.
    :param rpc_urls: Optional mapping of CAIP-2 network identifier to
        JSON-RPC endpoint URL. Entries are merged over the built-in
        :data:`~cdp_x402.core.constants.CDP_EVM_RPC_URLS` defaults — only
        supply the networks you want to override (e.g. to use a private
        Alchemy / Infura endpoint instead of the free-tier public RPC).
    :returns: An async before-payment-creation hook.
    """
    warn = on_warning or default_warn
    resolved_rpc_urls: dict[str, str] = (
        {**CDP_EVM_RPC_URLS, **rpc_urls} if rpc_urls else CDP_EVM_RPC_URLS
    )

    async def hook(context: Any) -> None:
        req = context.selected_requirements
        required = read_required_atomic(req)
        if required is None or required == 0:
            return None

        network: str = req.network if hasattr(req, "network") else ""
        asset: str = req.asset if hasattr(req, "asset") else ""
        normalized_asset = normalize_asset(asset)

        evm_network = _CAIP_TO_CDP_EVM_NETWORK.get(network)
        evm_rpc_url = resolved_rpc_urls.get(network)
        svm_network = _CAIP_TO_CDP_SVM_NETWORK.get(network)

        available: int
        address: str

        asset_label = asset if svm_network else normalized_asset

        try:
            if evm_network:
                address = evm_address
                available = await _get_evm_balance(
                    cdp_client,
                    address=evm_address,
                    network=evm_network,
                    asset=normalized_asset,
                )
            elif evm_rpc_url:
                if not normalized_asset.startswith("0x"):
                    return None
                address = evm_address
                available = await _read_onchain_erc20_balance(
                    evm_rpc_url,
                    token_address=normalized_asset,
                    wallet_address=evm_address,
                )
            elif svm_network:
                address = svm_address
                available = await _get_svm_balance(
                    cdp_client,
                    address=svm_address,
                    network=svm_network,
                    asset=asset,
                )
            else:
                return None
        except Exception as exc:
            warn(
                f"Pre-flight balance check failed for {network} ({asset_label}); "
                "proceeding without check.",
                exc,
            )
            return None

        if available >= required:
            return None

        raise InsufficientFundsError(
            required=required,
            available=available,
            asset=asset,
            network=network,
            address=address,
        )

    return hook
